### PR TITLE
Atomic updates in cassandra aspect dao

### LINKED ISSUE #5997: Atomic updates in CassandraAspectDao
## The problem

Updating an aspect in GMS translates into two operations in database: 
* Update the current latest aspect with a version number (meaning it's no longer latest);
* Insert a new aspect marked as latest. 

It is important to execute both operations atomically and in isolation to avoid inconsistent state.

GMS was initially designed to work with a relational database and the solution here is easy as you can begin and commit transactions at will.

With Cassandra backend, atomic/isolated execution of these two operations is possible by using Cassandra batches, but it is not currently used, meaning that GMS with a Cassandra backend doesn't have the protection against inconsistent state. We would very much like to change that.

## How things work now

`AspectDao` interface exposes a method `runInTransactionWithRetry` that allows execution of arbitrary code wrapped in a transaction. `EntityService` makes use of this method to ensure these insert+update operations happen in a single transaction. This behavior can't be matched in `CassandraAspectDao`.

## How we want things to work

`AspectDao` interface should be changed so that its clients can't rely on transaction controls that may or may not be available depending on the backing data store. It's probably best to move this update+insert complexity to `AspectDao` implementations to avoid leaking relational/cassandra implementation details to `EntityService`.

### PR DESCRIPTION
**Pull request by djordje-mijatovic**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

resolve #5997

### CODE REVIEW

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java [hunk 1]
```diff
@@ -526,17 +495,75 @@ public long saveLatestAspect(
     }
     // Save oldValue as the largest version + 1
     long largestVersion = ASPECT_LATEST_VERSION;
+    BatchStatement batch = BatchStatement.newInstance(BatchType.LOGGED);
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:498]: Does this statement need to be LOGGED? I thought `urn`+`aspectName` would always be on a single partition, thus avoiding the need to log (and its performance penalty).

I'm taking this as my inspiration as I have very little Cassandra knowledge: https://www.baeldung.com/java-cql-cassandra-batch#2-single-partition-batch

**Comment by CallumREdwards** [comment]
Nice! 

Unfortunately, a sad thing about Cassandra is that you can't mix LWT and non-LWT operations on the same table (see https://issues.apache.org/jira/browse/CASSANDRA-11000 and https://icircuit.net/cassandra-mixing-regular-writes-with-lwt-writes/3228). I can see that there are other mutations that occur on the aspect table without a lwt, which might cause some data corruption. Do we want to add lwt operations to the other mutations in the same PR?

**Comment by djordje-mijatovic** [comment]
> Do we want to add lwt operations to the other mutations in the same PR?

Definitely yes.

**Review by jjoyce0510 — APPROVED** [review]
This looks great to me! Once passes CI, I'll go ahead and merge

**Pull request merged**